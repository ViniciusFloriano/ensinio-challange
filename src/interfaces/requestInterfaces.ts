import { LanguagesProps } from './languageInterfaces';

export interface ItemProps {
  id: number;
  title: LanguagesProps;
  description: LanguagesProps;
}
