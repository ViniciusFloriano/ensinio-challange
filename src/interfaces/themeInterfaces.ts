export interface ThemeProps {
  colors: {
    white?: string;
    black?: string;
    gray?: string;
    lightGray?: string;
  };
  sizes: {
    xsmall: string | number;
    small: string | number;
    normal: string | number;
    large: string | number;
    xlarge: string | number;
    xxlarge: string | number;
  };
}
