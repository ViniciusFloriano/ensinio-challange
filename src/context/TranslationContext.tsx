import React, { createContext, useEffect, useState } from 'react';

import { LanguagesProps } from '../interfaces/languageInterfaces';
import { TranslationProps } from '../interfaces/requestInterfaces';
import { getTranslations } from '../services/requests';

type LanguagesType = keyof LanguagesProps;

interface TranslationContextProps {
  loading: boolean;
  error: boolean;
  currentLang: LanguagesType;
  setCurrentLang: (language: LanguagesType) => void;
  t: (title: string) => string;
  translationData: TranslationProps[];
}

const Context = createContext<TranslationContextProps>({
  loading: true,
  error: false,
  currentLang: 'pt',
  setCurrentLang: () => {},
  t: (title: string) => title,
  translationData: [],
});

function useTranslation() {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(true);
  const [translationData, setTranslationData] = useState<TranslationProps[]>(
    []
  );
  const [currentLang, setCurrentLangState] = useState<LanguagesType>('pt');

  useEffect(() => {
    getTranslations()
      .then((data) => {
        setTranslationData(data);
        setLoading(false);
      })
      .catch(() => {
        setError(true);
      });

    const languageStorage = localStorage.getItem('language');

    if (languageStorage) {
      setCurrentLangState(JSON.parse(languageStorage));
    } else {
      const navigatorLanguage = navigator.language;
      if (navigatorLanguage.includes('en')) {
        setCurrentLangState('en');
      } else if (navigatorLanguage.includes('pt')) {
        setCurrentLangState('pt');
      } else if (navigatorLanguage.includes('es')) {
        setCurrentLangState('es');
      }
    }
  }, []);

  function t(title: string) {
    const translatedText = translationData.find(
      (td) => td.title.toLowerCase() === title.toLowerCase()
    );

    if (!translatedText) {
      return 'Translation not find';
    }

    return translatedText.value[currentLang];
  }

  function setCurrentLang(language: LanguagesType) {
    localStorage.setItem('language', JSON.stringify(language));
    setCurrentLangState(language);
  }

  return { loading, error, currentLang, setCurrentLang, t, translationData };
}

const TranslationProvider: React.FC = ({ children }) => {
  const {
    loading,
    error,
    currentLang,
    setCurrentLang,
    t,
    translationData,
  } = useTranslation();

  return (
    <Context.Provider
      value={{
        loading,
        error,
        currentLang,
        setCurrentLang,
        t,
        translationData,
      }}
    >
      {children}
    </Context.Provider>
  );
};

export { Context as TranslationContext, TranslationProvider };
