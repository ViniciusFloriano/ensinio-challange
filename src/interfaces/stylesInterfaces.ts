import { ThemeProps } from './themeInterfaces';

export interface HeadingProps {
  color?: keyof ThemeProps['colors'];
  size?: keyof ThemeProps['sizes'];
  fontWeight?: 100 | 400 | 700;
  lineHeight?: string | number;
  level?: 1 | 2 | 3 | 4 | 5 | 6;
}
