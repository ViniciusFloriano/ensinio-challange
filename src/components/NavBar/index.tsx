import React, { useContext, useState } from 'react';

import LogoImage from '../../assets/icons/logo.svg';
import ProfileIcon from '../../assets/icons/profile';
import { TranslationContext } from '../../context/TranslationContext';
import Container from '../Container';
import NavItemsList, { NavItemPropsExtended } from '../NavItemsList/index';
import NavBarContainer from './styles';

const EmptyComponent: React.FC = () => <div />;

const NavBar: React.FC = () => {
  const { t, currentLang } = useContext(TranslationContext);
  const [navbarBgActive, setNavbarBgActive] = useState(false);

  const centerItems: NavItemPropsExtended[] = [
    { key: '01', text: t('Solutions'), dropdownComponent: EmptyComponent },
    { key: '02', text: t('Prices') },
    { key: '03', text: t('Academy') },
    { key: '04', text: t('Blog') },
    { key: '05', text: t('Contact') },
  ];

  const rightItems: NavItemPropsExtended[] = [
    { key: '11', leftIcon: ProfileIcon, text: t('Sing On') },
    { key: '12', text: t('Start now'), type: 'button' },
    {
      key: '13',
      dropdownComponent: EmptyComponent,
      text: currentLang.toUpperCase(),
    },
  ];

  function changeBackground() {
    if (window.scrollY >= 15) {
      setNavbarBgActive(true);
    } else {
      setNavbarBgActive(false);
    }
  }

  window.addEventListener('scroll', changeBackground);

  return (
    <NavBarContainer navbarBgActive={navbarBgActive}>
      <Container>
        <NavItemsList
          centerItems={centerItems}
          logo={LogoImage}
          rightItems={rightItems}
        />
      </Container>
    </NavBarContainer>
  );
};

export default NavBar;
