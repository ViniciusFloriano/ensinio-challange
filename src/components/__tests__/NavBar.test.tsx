import React from 'react';
import renderer from 'react-test-renderer';

import { cleanup } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { TranslationContext } from '../../context/TranslationContext';
import { LanguagesProps } from '../../interfaces/languageInterfaces';
import GlobalStyle from '../../styles/global';
import defaultTheme from '../../styles/themes';
import NavBar from '../NavBar';

it('renders correctly when there are no items', () => {
  const language: keyof LanguagesProps = 'pt';
  const translation = {
    loading: true,
    error: false,
    currentLang: language,
    setCurrentLang: () => {},
    t: (title: string) => title,
    translationData: [],
  };

  const tree = renderer
    .create(
      <TranslationContext.Provider value={translation}>
        <ThemeProvider theme={defaultTheme}>
          <NavBar />
          <GlobalStyle />
        </ThemeProvider>
      </TranslationContext.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
afterEach(cleanup);
