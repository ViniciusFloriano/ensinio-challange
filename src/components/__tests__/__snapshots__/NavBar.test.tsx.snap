// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<header
  className="sc-jrsJWt kPSpEw"
>
  <div
    className="sc-bdnxRM bRYKDy"
  >
    <div
      className="sc-gKAaRy bicxQW"
    >
      <img
        alt="logo"
        src="logo.svg"
      />
      <nav>
        <div
          id="nav-center-items"
        >
          <ul>
            <li>
              <a
                className="sc-eCApnc hBtWXp"
              >
                <span
                  className="sc-gtsrHT lkYkaj dropdown"
                  kind="span"
                  size="xxsmall"
                >
                  Solutions
                   
                  <svg
                    fill="#99F3F5"
                    height="8"
                    viewBox="0 0 10 8"
                    width="10"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.848 6.6432C5.45633 7.26987 4.54367 7.26987 4.152 6.6432L0.956249 1.53C0.539968 0.86395 1.01881 1.68514e-07 1.80425 2.37179e-07L8.19575 7.95942e-07C8.98119 8.64607e-07 9.46003 0.863952 9.04375 1.53L5.848 6.6432Z"
                    />
                  </svg>
                </span>
                <div
                  className="dropdown-content"
                >
                  <div
                    className="sc-pNWdM gUSWKd"
                  >
                    <h3
                      className="sc-iCoGMd joObkW"
                      color="linkPrimaryOpacity"
                      fontFamily="fontSecondary"
                      fontWeight={500}
                      letterSpacing="0.12rem"
                      size="xsmall"
                    >
                      Main Solutions
                    </h3>
                    <div>
                      <div
                        className="sc-fujyAs dZaOPG"
                      >
                        <img
                          alt="Create an Online School"
                          src="ead.svg"
                        />
                        <div>
                          <h1
                            className="sc-iCoGMd iSGkav"
                            color="textTertiary"
                            size="xsmall"
                          >
                            Create an Online School
                          </h1>
                          <p
                            className="sc-gtsrHT jwszLL"
                            color="textTertiaryOpacity"
                            fontWeight={400}
                            size="xsmall"
                          >
                            Lorem ipsum dolor sit amet
                          </p>
                        </div>
                      </div>
                      <div
                        className="sc-fujyAs dZaOPG"
                      >
                        <img
                          alt="Community and social network"
                          src="icon-app.svg"
                        />
                        <div>
                          <h1
                            className="sc-iCoGMd iSGkav"
                            color="textTertiary"
                            size="xsmall"
                          >
                            Community and social network
                          </h1>
                          <p
                            className="sc-gtsrHT jwszLL"
                            color="textTertiaryOpacity"
                            fontWeight={400}
                            size="xsmall"
                          >
                            Lorem ipsum dolor sit amet
                          </p>
                        </div>
                      </div>
                      <div
                        className="sc-fujyAs dZaOPG"
                      >
                        <img
                          alt="Gamification"
                          src="icon-gamification.svg"
                        />
                        <div>
                          <h1
                            className="sc-iCoGMd iSGkav"
                            color="textTertiary"
                            size="xsmall"
                          >
                            Gamification
                          </h1>
                          <p
                            className="sc-gtsrHT jwszLL"
                            color="textTertiaryOpacity"
                            fontWeight={400}
                            size="xsmall"
                          >
                            Lorem ipsum dolor sit amet
                          </p>
                        </div>
                      </div>
                      <div
                        className="sc-fujyAs dZaOPG"
                      >
                        <img
                          alt="Mobile app"
                          src="icon-social.svg"
                        />
                        <div>
                          <h1
                            className="sc-iCoGMd iSGkav"
                            color="textTertiary"
                            size="xsmall"
                          >
                            Mobile app
                          </h1>
                          <p
                            className="sc-gtsrHT jwszLL"
                            color="textTertiaryOpacity"
                            fontWeight={400}
                            size="xsmall"
                          >
                            Lorem ipsum dolor sit amet
                          </p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </a>
            </li>
            <li>
              <a
                className="sc-eCApnc hBtWXp"
              >
                Prices
              </a>
            </li>
            <li>
              <a
                className="sc-eCApnc hBtWXp"
              >
                Academy
              </a>
            </li>
            <li>
              <a
                className="sc-eCApnc hBtWXp"
              >
                Blog
              </a>
            </li>
            <li>
              <a
                className="sc-eCApnc hBtWXp"
              >
                Contact
              </a>
            </li>
          </ul>
        </div>
        <div
          className="splitter"
        >
          Â 
        </div>
        <div
          id="nav-right-items"
        >
          <ul>
            <li
              data-testid="nav-item-right-11"
            >
              <a
                className="sc-eCApnc hBtWXp"
              >
                <span
                  className="sc-gtsrHT lkYkaj"
                  id="icon"
                  kind="span"
                  size="xxsmall"
                >
                  <svg
                    fill="none"
                    height="20"
                    stroke="#99F3F5"
                    viewBox="0 0 20 20"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clipRule="evenodd"
                      d="M5 15.4866C6.3075 14.2157 8.07667 13.4341 10.0192 13.4341C11.9425 13.4341 13.6925 14.1999 15 15.4474C13.6925 16.7182 11.9233 17.4999 9.98083 17.4999C8.0575 17.4999 6.3075 16.7341 5 15.4866Z"
                      fillRule="evenodd"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                    />
                    <path
                      d="M3.3125 13.3542C2.80417 12.3425 2.5 11.21 2.5 10C2.5 5.855 5.855 2.5 10 2.5C14.145 2.5 17.5 5.855 17.5 10C17.5 11.2083 17.1975 12.3417 16.6858 13.3525"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                    />
                    <path
                      d="M11.7678 6.56573C12.7441 7.54204 12.7441 9.12495 11.7678 10.1013C10.7915 11.0776 9.20854 11.0776 8.23223 10.1013C7.25592 9.12495 7.25592 7.54204 8.23223 6.56573C9.20854 5.58942 10.7915 5.58942 11.7678 6.56573"
                      strokeLinecap="round"
                      strokeLinejoin="round"
                    />
                  </svg>
                   
                  Sing On
                </span>
              </a>
            </li>
            <li
              data-testid="nav-item-right-12"
            >
              <a
                className="sc-jSFjdj UepUh"
              >
                Start now
              </a>
            </li>
            <li
              data-testid="nav-item-right-13"
            >
              <a
                className="sc-eCApnc hBtWXp"
              >
                <span
                  className="sc-gtsrHT lkYkaj dropdown"
                  kind="span"
                  size="xxsmall"
                >
                  PT
                   
                  <svg
                    fill="#99F3F5"
                    height="8"
                    viewBox="0 0 10 8"
                    width="10"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.848 6.6432C5.45633 7.26987 4.54367 7.26987 4.152 6.6432L0.956249 1.53C0.539968 0.86395 1.01881 1.68514e-07 1.80425 2.37179e-07L8.19575 7.95942e-07C8.98119 8.64607e-07 9.46003 0.863952 9.04375 1.53L5.848 6.6432Z"
                    />
                  </svg>
                </span>
                <div
                  className="dropdown-content"
                >
                  <div
                    className="sc-hKFxyN kesLzR"
                  >
                    <div
                      aria-hidden="true"
                      checked={true}
                      className="sc-dlnjwi giMfzk"
                      onClick={[Function]}
                    >
                      <img
                        alt="PT"
                        src="brazil.svg"
                      />
                      <p
                        className="sc-gtsrHT kuwjhP"
                        color="textTertiary"
                        fontWeight={400}
                        size="small"
                      >
                        PT
                      </p>
                      <img
                        alt="PT"
                        src="check.svg"
                      />
                    </div>
                    <div
                      aria-hidden="true"
                      checked={false}
                      className="sc-dlnjwi jpqoUf"
                      onClick={[Function]}
                    >
                      <img
                        alt="EN"
                        src="usa.svg"
                      />
                      <p
                        className="sc-gtsrHT kuwjhP"
                        color="textTertiary"
                        fontWeight={400}
                        size="small"
                      >
                        EN
                      </p>
                    </div>
                    <div
                      aria-hidden="true"
                      checked={false}
                      className="sc-dlnjwi jpqoUf"
                      onClick={[Function]}
                    >
                      <img
                        alt="ES"
                        src="spain.svg"
                      />
                      <p
                        className="sc-gtsrHT kuwjhP"
                        color="textTertiary"
                        fontWeight={400}
                        size="small"
                      >
                        ES
                      </p>
                    </div>
                  </div>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </nav>
      <div
        aria-hidden="true"
        id="hamburguer-menu"
        onClick={[Function]}
      >
        <img
          alt="Menu"
          src="menu.svg"
        />
      </div>
    </div>
  </div>
</header>
`;
